### PR TITLE
Stream KTO log probabilities

### PR DESCRIPTION
**Pull request by kashif**
# What does this PR do?

Moves KTO's memory-efficient path onto `_ChunkedLogProbFunction`, including policy, reference, and KL log probabilities. The trainer's existing loss math now handles KTO, APO-zero, class weights, and precomputed references, and the duplicated KTO loss files are removed.

Part of #7063.

## Before submitting

- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/trl/blob/main/CONTRIBUTING.md#create-a-pull-request), Pull Request section?
- [x] Was this discussed/approved via a GitHub issue? Please add a link to it if that's the case.
- [x] Did you make sure to update the documentation with your changes?
- [x] Did you write any new necessary tests?

## AI writing disclosure

- [ ] No AI usage: the PR was written entirely by a human.
- [ ] AI-assisted: some parts were suggested or improved by AI, but the PR was written and reviewed by a human.
- [x] AI-generated: the PR was mostly or fully generated by an AI tool.

## Who can review?

@qgallouedec

---

> [!NOTE]
> **Medium Risk**
> Changes core KTO training and gradient paths when `use_liger_kernel=True`, including distributed (ZeRO-3/FSDP2) reference forwards; behavior is heavily tested but touches alignment loss numerics and memory-efficient execution.
> 
> **Overview**
> **Replaces the dedicated Liger fused KTO loss with streamed log-probability computation** via `_ChunkedLogProbFunction`, so `use_liger_kernel=True` shares the same KTO loss math as the standard path (including `apo_zero_unpaired`, class weights, and precomputed reference log-probs).
> 
> The trainer adds `_get_per_token_logps_and_entropies` for policy, reference, and KL forwards without full `[N, V]` logits; **FSDP2 is supported** by gathering sharded `lm_head` DTensors and routing distributed forwards through `_forward_redirection`. Chunked logprob autograd gains **token-wise chunking**, autocast-aware dtypes, and safer mixed-precision backward.
> 
> **Removes** `FusedLinearKTOLoss` and related loss modules plus their parity tests. Docs note omitted `logits/*` metrics on the chunked path and narrowed `use_liger_kernel` PEFT/MoE constraints. Distributed KTO+Liger **no longer xfail on FSDP2**; new trainer/utils tests cover loss parity, precompute ref logps, and chunked behavior.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit 12932320ddbaba99e39238566efbba73b6049cd9. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

File: trl/trainer/kto_trainer.py [hunk 1]
```diff
@@ -1627,24 +1546,25 @@ def _compute_loss(self, model, inputs, return_outputs):
...
-        total_rejected_tokens = rejected_mask.sum()
-        total_chosen_logits = self.accelerator.gather_for_metrics(total_chosen_logits).sum().item()
-        total_chosen_tokens = self.accelerator.gather_for_metrics(total_chosen_tokens).sum().item()
-        total_rejected_logits = self.accelerator.gather_for_metrics(total_rejected_logits).sum().item()
-        total_rejected_tokens = self.accelerator.gather_for_metrics(total_rejected_tokens).sum().item()
-        if total_chosen_tokens > 0:
-            self._metrics[mode]["logits/chosen"].append(total_chosen_logits / total_chosen_tokens)
-        if total_rejected_tokens > 0:
-            self._metrics[mode]["logits/rejected"].append(total_rejected_logits / total_rejected_tokens)
+        if not self.use_liger_kernel:
```
[the remark this answers is not in the document]
Comment (kashif) [reply trl/trainer/kto_trainer.py:1576]: Documented in 3944b6cf. Keeping these two metrics would require either another full LM-head projection or expanding the shared chunked primitive beyond the per-token log-probs/entropy needed by the trainer. The docs now state that they are omitted when the chunked path is enabled.

File: trl/trainer/kto_trainer.py [hunk 2]
```diff
@@ -1260,33 +1246,92 @@ def add_ref_logps(batch, indices):
...
+                )
+                if self.calculate_KL:
+                    KL_kwargs = {
+                        "input_ids": inputs["KL_input_ids"],
+                        "attention_mask": inputs["KL_attention_mask"],
+                        "use_cache": False,
+                    }
+                    KL_per_token_logps, _, _ = self._get_per_token_logps_and_entropies(
+                        model, KL_kwargs, inputs["KL_input_ids"], inputs["KL_completion_mask"]
+                    )
```
[the remark this answers is not in the document]
Comment (kashif) [reply trl/trainer/kto_trainer.py:1345]: Fixed in 3944b6cf. The chunked reference path now redirects wrapped models through the DeepSpeed/FSDP wrapper before calling the unwrapped backbone. The regression test verifies that precomputation cannot bypass the wrapper.

**Comment by kashif** [comment]
@codex review

File: trl/trainer/kto_trainer.py [hunk 3]
```diff
@@ -1260,33 +1246,103 @@ def add_ref_logps(batch, indices):
...
+        hidden_states = outputs.last_hidden_state[:, :-1]
+        labels = input_ids[:, 1:]
+        mask = completion_mask[:, 1:].bool()
+        hidden_states = hidden_states[mask]
+        labels = labels[mask]
+
+        lm_head = inner_model.get_output_embeddings()
+        text_config = inner_model.config.get_text_config()
+        final_logit_softcapping = getattr(text_config, "final_logit_softcapping", None)
+        logit_scale = getattr(text_config, "logit_scale", 1.0)
```
[the remark this answers is not in the document]
Comment (kashif) [reply trl/trainer/kto_trainer.py]: Fixed in dbbd3628. KTO now normalizes `None`, honors `output_multiplier`, and has a regression that checks the projected log-probs. The same pass also retains bounded token chunks and materializes FSDP2 head DTensors, matching the GRPO/SFT paths.

**Comment by kashif** [comment]
@codex review

File: trl/trainer/kto_trainer.py [hunk 4]
```diff
@@ -1260,33 +1247,129 @@ def add_ref_logps(batch, indices):
...
+                    labels_chunk,
+                    1.0,
+                    _CHUNKED_LOGPROB_CHUNK_SIZE,
+                    final_logit_softcapping,
+                    logit_scale,
+                )
+                logps_chunks.append(logps_chunk)
+                entropy_chunks.append(entropy_chunk)
+        logps = torch.cat(logps_chunks)
+        entropies = torch.cat(entropy_chunks)
```
[the remark this answers is not in the document]
Comment (kashif) [reply trl/trainer/kto_trainer.py]: Fixed in df656530. Token tiling now lives inside the single shared autograd function, which preserves the empty-input behavior without an outer `cat`.

File: trl/trainer/kto_trainer.py [hunk 5]
```diff
@@ -1260,33 +1247,129 @@ def add_ref_logps(batch, indices):
...
+            lm_head_weight = lm_head_weight.full_tensor()
+            if lm_head_bias is not None:
+                lm_head_bias = lm_head_bias.full_tensor()
+
+        logps_chunks = []
+        entropy_chunks = []
+        with maybe_gather_lm_head_ctx(lm_head_weight, lm_head_bias):
+            for hidden_chunk, labels_chunk in zip(
+                hidden_states.split(_CHUNKED_LOGPROB_TOKEN_CHUNK_SIZE),
+                labels.split(_CHUNKED_LOGPROB_TOKEN_CHUNK_SIZE),
```
[the remark this answers is not in the document]
Comment (kashif) [reply trl/trainer/kto_trainer.py]: Fixed in df656530. There is now one `_ChunkedLogProbFunction` node and one backward gather context per rank; token tiling happens inside that context, so valid-token counts cannot change collective counts.

File: trl/trainer/kto_trainer.py [hunk 6]
```diff
@@ -1260,33 +1247,129 @@ def add_ref_logps(batch, indices):
...
+
+        lm_head = inner_model.get_output_embeddings()
+        text_config = inner_model.config.get_text_config()
+        final_logit_softcapping = getattr(text_config, "final_logit_softcapping", None)
+        # `logit_scale` is None on models that don't scale (e.g. MPT); read that as unscaled (1.0). A real 0.0 is
+        # kept as-is. Muse Glimmer applies the same pre-softcap multiplier as `output_multiplier`.
+        logit_scale = getattr(text_config, "logit_scale", None)
+        if logit_scale is None:
+            logit_scale = getattr(text_config, "output_multiplier", None)
+        logit_scale = 1.0 if logit_scale is None else logit_scale
```
[the remark this answers is not in the document]
Comment (kashif) [reply trl/trainer/kto_trainer.py:1274]: Fixed in df656530. The shared `patch_chunked_lm_head` path now uses the same `None` normalization and `output_multiplier` fallback. GRPO carries the identical shared change.

**Comment by kashif** [comment]
@codex review

File: trl/trainer/utils.py [hunk 7]
```diff
@@ -1415,58 +1437,71 @@ def backward(ctx, grad_logprobs: torch.Tensor | None, grad_entropy: torch.Tensor
...
+                        )
+
+                    grad_logits = grad_logits * inv_t
+                    if final_logit_softcapping is not None:
+                        grad_logits.mul_(1 - tanh_scaled.pow(2))
+
+                    grad_logits = grad_logits * logit_scale
+
+                    grad_hidden[token_start:token_end].add_(grad_logits @ w_chunk.float())
+                    grad_weight[start:end].add_(grad_logits.t() @ hidden_chunk.float())
```
[the remark this answers is not in the document]
Comment (kashif) [reply trl/trainer/utils.py]: Fixed in ed953c90. Backward now keeps the autocast-rounded hidden chunk used for projection recomputation and uses those same values for the LM-head weight gradient. The regression uses two fp32 hidden tensors that become identical in bf16; it verifies both projected log-probs and weight gradients are identical, so discarded fp32 bits cannot affect backward.

**Comment by kashif** [comment]
@codex review

**Comment by qgallouedec** [comment]
Thanks, same shape as #7077!

**Review by qgallouedec — APPROVED** [review]
